Addressing issue #30 about NIO FileSystem lifecycle

stefanofornari opened the pull request:
*Issue #, if available:*
#30 

*Description of changes:*
As per the discussion on #30 we want to implement NIO FileSystem semantic, which is lead by the methods newFileSystem() and getFileSystem(); the former shall return a new instance or throw an exception, while the latter should return an existing instance. This means that S3FileSystem instances shall be cached (as opposed to the current implementation that caches AWSClients).
Another change introduced and discussed in other thread is the use of S3ClientProvider at S3FileSystem level in order to better decouple file system logic from the aws client creation logic (discussed in #50). This obsoletes IMHO the methods in S3FileSystemProvider that receives a AWSAsyncClient, which I would remove in future version to keep S3FileSystemProvider cleaner and simpler.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

PS: as often gradle complains, this time about javadoc I think :( it builds locally with both gradle and maven ...

stefanofornari commented:
Hi @markjschreiber, do you mind looking at this one? It's probably the last one with remarkable architectural changes to the current codebase. After that we should be moving smoothly with the S3X provider.

markjschreiber reviewed (changes requested):
Thanks for putting in this PR. I have some changes requested but I think we can get there.

stefanofornari commented:
thanks! I am going to work on it in the next days

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3ClientProvider.java:
```diff
@@ -0,0 +1,288 @@
+            .maxBackoffTime(Duration.ofSeconds(5L))
+            .build();
+
+    final RetryCondition retryCondition;
```
No particular reason, it is coming from the original implementation. I guess it comes from being DEFAULT clients static instances. If we want to improve it, I would suggest to keep it for later.

markjschreiber replied:
It might actually be good to move the back off strategy and retry strategy objects out to a utility class and make them public static final objects to promote reuse?

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3ClientProvider.java:
```diff
@@ -0,0 +1,288 @@
+     * @return an asynchronous S3 client appropriate for the region of the named bucket
+     */
+    protected S3AsyncClient generateAsyncClient(String bucketName){
+        return generateAsyncClient(bucketName, DEFAULT_CLIENT);
```
This is coming as well from the old code in S3ClientStore. I assumed it was wanted because used primarily to locate a bucket, but let me know if we need to change it. I would do it anyway post merge to make the merge easier.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3ClientProvider.java:
```diff
@@ -0,0 +1,288 @@
+                            sdkHttpResponse()
+                            .firstMatchingHeader("x-amz-bucket-region")
+                            .orElseThrow(() -> new NoSuchElementException("Head Bucket Response doesn't include the header 'x-amz-bucket-region'")));
+                } catch (S3Exception e2) {
```
I do not find it that inelegant, actually. It could probably be refactored with some lambdas to be invoked on errors, but I would keep it as is for now.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystem.java:
```diff
@@ -89,7 +87,36 @@ public FileSystemProvider provider() {
+     *
+     * @return the client provider
+     */
+    public S3ClientProvider clientProvider() {
```
I like this syntax more as it is more concise and the same used by java records. I would keep it as it is, unless you have a strong opinion. Feel free to reopen the conversation if that is the case.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -108,10 +110,48 @@ public String getScheme() {
+        }
+        if (uri.getScheme() == null) {
+            throw new IllegalArgumentException(
+                String.format("invalid uri '%s', please provide an uri as s3://[key:secret@][host:port]/bucket", uri.toString())
```
yes, sorry, a leftover of the new scheme implementation, that as we have decided will go in a different provider; no need to mention it here for now.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -149,7 +189,43 @@ public FileSystem newFileSystem(URI uri, Map<String, ?> env) {
     @Override
     public S3FileSystem getFileSystem(URI uri) {
-        return new S3FileSystem(uri, this);
+        return getFileSystem(uri, false);
```
It is already in the throws, isn't it?

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystem.java:
```diff
@@ -89,7 +87,36 @@ public FileSystemProvider provider() {
+    /**
+     * @return the S3Client associated with this FileSystem
+     */
+    public S3AsyncClient client() {
```
as above

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -426,60 +523,31 @@ public void delete(Path path) throws IOException {
         }
     }
 
-    protected void delete(S3AsyncClient s3Client, Path path) throws ExecutionException, InterruptedException {
```
yep, this is a mistake, I will have it back.

stefanofornari replied:
it was actually like createDirectory: the method is there (maybe I moved it) and is commented.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -493,8 +561,56 @@ private static List<List<ObjectIdentifier>> getContainedObjectBatches(S3AsyncCli
     @Override
     public void copy(Path source, Path target, CopyOption... options) throws IOException {
+        //
+        // TODO: source and target can belong to any file system, we can not
```
agreed, it was a old comment, I think I also read that copy between different sources may not be supported. I will try it out. For now I would keep the comment (maybe rephrasing it so that we have a TODO).

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -231,18 +307,24 @@ protected SeekableByteChannel newByteChannel(S3AsyncClient client, Path path, Se
     }
 
+    /**
+     * @deprecate -- TODO how to address the deprecation
```
As mentioned somewhere, like all other methods that get a S3Client for testing purposing, I am deprecating the method because it is not needed any more and the preferred replacement is by the mean of the client provider. The reason why I would like to get rid of these methods in the future, in addition to make the interface of S3ClientProvider cleaner, is for maintenability purposes (i.e. once we have a protected method it becomes untouchable for backward compatibility).

I forgot to comment the method accordingly to the other:

```
/**
     *
     * @deprecated in favour of using a proper S3ClientProvider in S3FileSystem.
     *             For instance, instead of the following code:
     *             <pre>
     *             S3FileSystemProvider p = ...;
     *             S3Path dir = ...;
     *             S3AsyncClient s3 = ...;
     *             DirectoryStream filter = ...;
     *
     *             p.newDirectoryStream(s3, path, filter);
     *             </pre>
     *             something equivalent to the below should be used:
     *             <pre>
     *             S3FileSystemProvider p = new MyFileSystemProvider(new MyClientProvider());
     *             S3Path dir = ...;
     *
     *             p.newDirectoryStreampath, filter);
     *             </pre>
     */
```

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -254,37 +336,12 @@ protected SeekableByteChannel newByteChannel(S3AsyncClient client, Path path, Se
-     * @throws ExecutionException if the client async call(s) are interrupted or time out.
-     * @throws InterruptedException if the client async call(s) are interrupted or time out.
-     */
-    protected DirectoryStream<Path> newDirectoryStream(S3AsyncClient s3Client, Path dir, DirectoryStream.Filter<? super Path> filter) throws ExecutionException, InterruptedException {
```
maybe I moved the method around... this method is still there.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -323,6 +380,9 @@ public Iterator<Path> iterator() {
      * @return an iterator for {@code Path}s constructed from the {@code ListObjectsV2Publisher}s responses.
      */
+    //
+    // TODO: shall this be private?
```
this is covered in #129. I think we have not released it yet. maybe you can take care of it?

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -371,8 +431,25 @@ protected Iterator<Path> pathIteratorForPublisher(
         try {
-            this.createDirectory(null, dir, attrs);
+            String directoryKey = ((S3Path)dir).toRealPath(NOFOLLOW_LINKS).getKey();
+            if (!directoryKey.endsWith(S3Path.PATH_SEPARATOR) && !directoryKey.isEmpty()) {
```
It is like in the original code, but you are right. I have created #134 so that we can fix it after the merge.

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/e831e674425fd62d48ed4464299d5461e1badad3/src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java#L343-L346

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -381,32 +458,29 @@ public void createDirectory(Path dir, FileAttribute<?>... attrs) throws IOExcept
+     *             </pre>
+     */
+     @Deprecated
+     protected void createDirectory(S3AsyncClient s3Client, Path dir, FileAttribute<?>... attrs) throws IOException, ExecutionException, InterruptedException {
```
same comment as above.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -877,4 +1032,45 @@ protected Map<String, Object> readAttributes(S3AsyncClient client, Path path, St
+        return keys;
+    }
+
+    private String getFileSystemKey(URI uri) {
```
yep, this is a leftover of the support for 3rd party services. I introduced the same logic as in S3FileSystem and added the comment:
```

//       this duplicates the logic in S3FileSystem and assumes the key
//       is the bucket name; we need to move the source of truth in
//       S3FileSystemProvider or a separate class and pass it to the
//       file system constructor.
```
Let's address it in the refactory needed to support the other services.

stefanofornari commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -701,7 +805,45 @@ public FileStore getFileStore(Path path) {
+                if (httpResponse.isSuccessful()) return;
+
+                //
+                // TODO: it looks like if status is not SUCCESS an ExecutionException
```
Just removing the code breaks a few test that use the mocks to simulate the status code, but that do not actually simulate the exception. I would leave it as is for now.

markjschreiber replied:
It maybe that these two if statements need to be moved into the catch as a way to determine the type of exception that occurred.

stefanofornari commented:
@markjschreiber I should be done with your comments, thanks a lot for your detailed review. I left open just one comment that was not clear to me. Let me know if I can do anything else.

markjschreiber commented:
Hi, thanks for this update. I will be out of the office for the next week and with very limited connectivity. I will attempt to provide reviews as I am able.

markjschreiber commented on src/main/java/software/amazon/nio/spi/s3/S3ClientStore.java:
```diff
@@ -44,71 +19,26 @@ public class S3ClientStore {
      * Get the ClientStore instance
      * @return a singleton
      */
+    @Deprecated
```
Can you update the Java docs with the new pattern that developers would use instead of this deprecated method?

markjschreiber commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystem.java:
```diff
@@ -30,9 +21,13 @@
 /**
  * A Java NIO FileSystem for an S3 bucket as seen through the lens of the AWS Principal calling the class.
+ *
+ * TODO: provide constructrs with configuration settings like endpoint, bucket
```
Typo. Constructors

markjschreiber commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -323,6 +402,9 @@ public Iterator<Path> iterator() {
      * @return an iterator for {@code Path}s constructed from the {@code ListObjectsV2Publisher}s responses.
      */
+    //
+    // TODO: shall this be private?
```
An alternative way to make this testable would be to refactor it out of this class and into a helper/ utility class as a public method. That will avoid this class being polluted with protected method that isn't central to the function of the class.

stefanofornari replied:
I actually found the proper (I think) solution: cover the cases by using newDirectoryStream and getting the iterator from it:

```
Iterator<Path> pathIterator =
            provider.newDirectoryStream(Paths.get(URI.create(pathUri+"/")), filter).iterator();
```

HOWEVER, I was not able to make the test on pagination to work. I think I have done it quite right, but I miss how pagination is controlled. Can you please have a look at it (method pathIteratorForPublisher_withPagination). For now I disabled it and added a TODO.

markjschreiber reviewed (commented):
Looking good. A couple of minor comments and some suggestions on possible refactoring of some of the protected/ utility pieces.

stefanofornari commented:
> Hi, thanks for this update. I will be out of the office for the next week and with very limited connectivity. I will attempt to provide reviews as I am able.

have fun!

stefanofornari commented:
> > Hi, thanks for this update. I will be out of the office for the next week and with very limited connectivity. I will attempt to provide reviews as I am able.
> 
> have fun!

Hey @markjschreiber , I hope you had a good time! can we resume this work? The PR should be ready to be merged IMHO, but I am indeed available for updates.

markjschreiber commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystem.java:
```diff
@@ -42,17 +37,21 @@ public class S3FileSystem extends FileSystem {
     public static final String BASIC_FILE_ATTRIBUTE_VIEW = "basic";
 
-    private final String bucketName;
+    protected S3ClientProvider clientProvider;
```
This can probably be private because you have public methods to access and set the client?

markjschreiber commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -493,8 +579,57 @@ private static List<List<ObjectIdentifier>> getContainedObjectBatches(S3AsyncCli
+                                            .destinationBucket(resolvedS3TargetPath.bucketName())
+                                            .destinationKey(resolvedS3TargetPath.getKey())
+                                            .build())
+                                    .build()).completionFuture().join();
```
It is usually better to use `get(timeOut, unit)` otherwise there is a risk that the method will hang indefinitely and never return. The `get` with time out has the advantage that the method will not wait forever but instead throw an exception that the user can catch and recover from if they want to.

markjschreiber commented on src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java:
```diff
@@ -710,58 +871,36 @@ public void checkAccess(Path path, AccessMode... modes) throws IOException {
 
     /**
-     * Returns a file attribute view of a given type. This method works in
+     * Returns a file attribute view of a given type.This method works in
```
Typo removed space after "."

markjschreiber reviewed (changes requested):
Apologies for the delay in the review. A few small changes requested but it should be quick to fix them.

stefanofornari commented:
I should have done with everything. let me know.

markjschreiber commented on src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java:
```diff
@@ -125,16 +237,17 @@ public void newDirectoryStream() throws ExecutionException, InterruptedException
-     */
     @Test
-    public void pathIteratorForPublisher() {
+    @Disabled
```
I'm not comfortable disabling this. If the pagination is not working then we have a regression and should investigate.

stefanofornari replied:
I agree 100%. As per comment https://github.com/awslabs/aws-java-nio-spi-for-s3/pull/133#discussion_r1285169401 I need your help. I think I found a clean solution to make the iterator private, but I could not make it work the test. I believe it is my lack of familiarity with mocking S3 objects because the logic is correct:

`provider.newDirectoryStream(Paths.get(URI.create(pathUri+"/")), filter).iterator()`

DirectoryStream already provides a way to get the iterator. Once you get that, the test should work exactly like before. Can you please have a look and help me out with that?

markjschreiber replied:
The challenge with making it private is that you can't provide it with an S3Client which ultimately means you can't provide it with a mock S3Client. So the Directory iterator you get will be using a regular S3Client and will try to call the S3 service which will fail. 

One way you could do it is to make a separate protected or even public ListObjectsV2Provider and create one in the constructor of the appropriate class as a private variable.

stefanofornari replied:
I do not think so, @markjschreiber. If you look run the test, what fails is the last test only, which means to me if is using the mock passed in in creating the publisher

https://github.com/stefanofornari/aws-java-nio-spi-for-s3/blob/896d1e812405a23bcfb007f134e1e7474bc2b45d/src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java#L251 

Plus, the test just before 

https://github.com/stefanofornari/aws-java-nio-spi-for-s3/blob/896d1e812405a23bcfb007f134e1e7474bc2b45d/src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java#L225C24-L225C24

Which basically the same, but it does not paginate, works, therefore, no it can not be that. The thing is that pagination is not triggered and maybe it was not before neither. I tend to believe it is in the way the mock is prepared. How can we trigger pagination by a proper client response?

stefanofornari replied:
I will keep investigating it, but since you developed that piece, I am pretty sure that you can spot the issue quickly if you give it a run.

stefanofornari replied:
Ok, I've got it... it was actually a bug

The problem was that in here:

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/98e2f9da844b12443b1841373a0643a4e56c576a/src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java#L348
p is a string containing the full path like s3://foo/baa/, while finalDirName contains only the path 'baa/' because it comes from:

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/98e2f9da844b12443b1841373a0643a4e56c576a/src/main/java/software/amazon/nio/spi/s3/S3FileSystemProvider.java#L303

where finalDirName is the result of s3Path.toAbsolutePath().getKey(); and the two never match.

The reason why your test was passing is because you gave a matching string in your case (you pass in pathUri+"/" as string argument, which turns into "s3://foo/baa/"), but this is not what production code does:

https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/98e2f9da844b12443b1841373a0643a4e56c576a/src/test/java/software/amazon/nio/spi/s3/S3FileSystemProviderTest.java#L150-L154

This is also why I avoid using mocks as much as I can ;)

I am updating the PR with a possible fix, please let me know if it is fine,

markjschreiber reviewed (commented):
Why is the pagination test disabled?

stefanofornari commented:
sorry, I closed it by mistake...

markjschreiber commented:
Re-opening

markjschreiber commented:
I just did some quick checks with the applications in `examples`. None of them currently work so something has broken in this PR for (at least) S3 native buckets/ URIs.

stefanofornari commented:
uh... unexpected : ) can you give me a bucket I can try with?

stefanofornari commented:
do you get the NPE: Exception in thread "main" java.lang.NullPointerException: Cannot invoke "software.amazon.nio.spi.s3.S3ClientProvider.generateAsyncClient(String) ?

stefanofornari commented:
yes, there is a bug here I believe. I am working on it

markjschreiber commented:
Yes, I get a null pointer exception for `main`

The other two example apps give "no such filesystem" errors.  For testing purposes any of the buckets listed in the "registry of open data" are usually OK. I tend to use `s3://broad-references/` or some text file therein.

stefanofornari commented:
ok, I fixed the issue. I introduced couple of regressions not captured by unit tests. I tested it with sample applications and it works now.

markjschreiber commented:
> ok, I fixed the issue. I introduced couple of regressions not captured by unit tests. I tested it with sample applications and it works now.

Nice. If you haven't already can you open an issue(s) that outline the code parts/ scenarios where we need to add unit tests to spot similar issues in the future? It would be better to rely on automated testing rather than manually running the example apps.

stefanofornari commented:
> > ok, I fixed the issue. I introduced couple of regressions not captured by unit tests. I tested it with sample applications and it works now.
> 
> Nice. If you haven't already can you open an issue(s) that outline the code parts/ scenarios where we need to add unit tests to spot similar issues in the future? It would be better to rely on automated testing rather than manually running the example apps.

#165
